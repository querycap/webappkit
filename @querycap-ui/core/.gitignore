!macro/index.js

!jsx-runtime/*.js