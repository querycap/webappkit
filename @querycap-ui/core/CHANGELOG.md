# Change Log

## 0.11.8

### Patch Changes

- fix exports

## 0.11.7

### Patch Changes

- 2b049a7: deps updates
- Updated dependencies [2b049a7]
  - @querycap-ui/core.macro@0.4.5

## 0.11.6

### Patch Changes

- release fix
- Updated dependencies [undefined]
  - @querycap-ui/core.macro@0.4.4
  - @querycap-ui/css-aliases@0.3.4

## 0.11.5

### Patch Changes

- release fix
- Updated dependencies [undefined]
  - @querycap-ui/core.macro@0.4.3
  - @querycap-ui/css-aliases@0.3.3

## 0.11.4

### Patch Changes

- 36cbb36: deps updates
- Updated dependencies [36cbb36]
  - @querycap-ui/core.macro@0.4.2
  - @querycap-ui/css-aliases@0.3.2

## 0.11.3

### Patch Changes

- first release by changesets
- Updated dependencies [undefined]
  - @querycap-ui/core.macro@0.4.1
  - @querycap-ui/css-aliases@0.3.1

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.11.2](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.11.1...@querycap-ui/core@0.11.2) (2021-04-25)

**Note:** Version bump only for package @querycap-ui/core

## [0.11.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.11.0...@querycap-ui/core@0.11.1) (2021-04-01)

**Note:** Version bump only for package @querycap-ui/core

# [0.11.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.10...@querycap-ui/core@0.11.0) (2021-03-24)

### Features

- drop yarn and use pnpm ([f7e7394](https://github.com/querycap/webappkit/commit/f7e7394e1531ffb96ecb3e393e8131451f3e1d9f))

## [0.10.10](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.9...@querycap-ui/core@0.10.10) (2021-03-23)

**Note:** Version bump only for package @querycap-ui/core

## [0.10.9](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.8...@querycap-ui/core@0.10.9) (2021-03-10)

**Note:** Version bump only for package @querycap-ui/core

## [0.10.8](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.7...@querycap-ui/core@0.10.8) (2021-02-05)

**Note:** Version bump only for package @querycap-ui/core

## [0.10.7](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.6...@querycap-ui/core@0.10.7) (2021-02-02)

**Note:** Version bump only for package @querycap-ui/core

## [0.10.6](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.5...@querycap-ui/core@0.10.6) (2020-12-21)

**Note:** Version bump only for package @querycap-ui/core

## [0.10.5](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.4...@querycap-ui/core@0.10.5) (2020-12-10)

**Note:** Version bump only for package @querycap-ui/core

## [0.10.4](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.3...@querycap-ui/core@0.10.4) (2020-12-09)

**Note:** Version bump only for package @querycap-ui/core

## [0.10.3](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.2...@querycap-ui/core@0.10.3) (2020-12-01)

**Note:** Version bump only for package @querycap-ui/core

## [0.10.2](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.1...@querycap-ui/core@0.10.2) (2020-11-23)

**Note:** Version bump only for package @querycap-ui/core

## [0.10.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.10.0...@querycap-ui/core@0.10.1) (2020-11-16)

**Note:** Version bump only for package @querycap-ui/core

# [0.10.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.9.5...@querycap-ui/core@0.10.0) (2020-11-16)

### Features

- **@querycap-ui/\*:** upgrade to [@emotion](https://github.com/emotion) to 11 ([5653bb6](https://github.com/querycap/webappkit/commit/5653bb63579fd592382fa4dd2ee709a838f6e944))

## [0.9.5](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.9.4...@querycap-ui/core@0.9.5) (2020-11-10)

**Note:** Version bump only for package @querycap-ui/core

## [0.9.4](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.9.3...@querycap-ui/core@0.9.4) (2020-11-02)

**Note:** Version bump only for package @querycap-ui/core

## [0.9.3](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.9.2...@querycap-ui/core@0.9.3) (2020-10-29)

**Note:** Version bump only for package @querycap-ui/core

## [0.9.2](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.9.0...@querycap-ui/core@0.9.2) (2020-10-15)

**Note:** Version bump only for package @querycap-ui/core

# [0.9.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.17...@querycap-ui/core@0.9.0) (2020-10-15)

### Features

- bump jsx-runtime ([453f52b](https://github.com/querycap/webappkit/commit/453f52b4a7b0e0f987de76da08c9bbb4d39802f8))

## [0.8.17](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.16...@querycap-ui/core@0.8.17) (2020-10-14)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.16](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.15...@querycap-ui/core@0.8.16) (2020-10-14)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.15](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.14...@querycap-ui/core@0.8.15) (2020-10-12)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.14](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.13...@querycap-ui/core@0.8.14) (2020-10-12)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.13](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.12...@querycap-ui/core@0.8.13) (2020-09-21)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.12](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.11...@querycap-ui/core@0.8.12) (2020-09-07)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.11](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.10...@querycap-ui/core@0.8.11) (2020-08-22)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.10](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.9...@querycap-ui/core@0.8.10) (2020-08-17)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.9](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.8...@querycap-ui/core@0.8.9) (2020-08-13)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.8](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.7...@querycap-ui/core@0.8.8) (2020-08-07)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.7](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.6...@querycap-ui/core@0.8.7) (2020-08-04)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.6](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.5...@querycap-ui/core@0.8.6) (2020-07-14)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.5](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.4...@querycap-ui/core@0.8.5) (2020-06-19)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.4](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.3...@querycap-ui/core@0.8.4) (2020-05-22)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.3](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.2...@querycap-ui/core@0.8.3) (2020-05-19)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.2](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.1...@querycap-ui/core@0.8.2) (2020-05-15)

**Note:** Version bump only for package @querycap-ui/core

## [0.8.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.8.0...@querycap-ui/core@0.8.1) (2020-05-04)

**Note:** Version bump only for package @querycap-ui/core

# [0.8.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.10...@querycap-ui/core@0.8.0) (2020-04-23)

### Features

- **ui/core:** drop theme space ([4aa3ac3](https://github.com/querycap/webappkit/commit/4aa3ac38d3dadcb124b83ac0d8e101213f14058a))

## [0.7.10](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.9...@querycap-ui/core@0.7.10) (2020-04-14)

### Bug Fixes

- **ui/core:** lineHeight in state ([8a46840](https://github.com/querycap/webappkit/commit/8a46840a4db2be1bc0afdc67bad5325e7bc9bcf5))

## [0.7.9](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.8...@querycap-ui/core@0.7.9) (2020-04-14)

**Note:** Version bump only for package @querycap-ui/core

## [0.7.8](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.7...@querycap-ui/core@0.7.8) (2020-04-13)

**Note:** Version bump only for package @querycap-ui/core

## [0.7.7](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.6...@querycap-ui/core@0.7.7) (2020-04-10)

**Note:** Version bump only for package @querycap-ui/core

## [0.7.6](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.5...@querycap-ui/core@0.7.6) (2020-04-10)

**Note:** Version bump only for package @querycap-ui/core

## [0.7.5](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.4...@querycap-ui/core@0.7.5) (2020-04-10)

**Note:** Version bump only for package @querycap-ui/core

## [0.7.4](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.3...@querycap-ui/core@0.7.4) (2020-04-03)

**Note:** Version bump only for package @querycap-ui/core

## [0.7.3](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.2...@querycap-ui/core@0.7.3) (2020-04-02)

**Note:** Version bump only for package @querycap-ui/core

## [0.7.2](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.1...@querycap-ui/core@0.7.2) (2020-04-02)

**Note:** Version bump only for package @querycap-ui/core

## [0.7.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.7.0...@querycap-ui/core@0.7.1) (2020-04-02)

**Note:** Version bump only for package @querycap-ui/core

# [0.7.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.6.4...@querycap-ui/core@0.7.0) (2020-04-02)

### Features

- **blocks:** mv Menu into blocks ([a0f538b](https://github.com/querycap/webappkit/commit/a0f538b36ad20c32b4df4b554a6a092891355f48))

## [0.6.4](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.6.3...@querycap-ui/core@0.6.4) (2020-04-01)

**Note:** Version bump only for package @querycap-ui/core

## [0.6.3](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.6.2...@querycap-ui/core@0.6.3) (2020-03-31)

**Note:** Version bump only for package @querycap-ui/core

## [0.6.2](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.6.1...@querycap-ui/core@0.6.2) (2020-03-31)

### Bug Fixes

- re-exports fix ([dc91ec7](https://github.com/querycap/webappkit/commit/dc91ec7983130355667eca8ce24c6dc5b0c4619e))

## [0.6.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.6.0...@querycap-ui/core@0.6.1) (2020-03-31)

### Bug Fixes

- **ui/core:** fix macro in node.js ([102af93](https://github.com/querycap/webappkit/commit/102af9372adae55c61f45221c1096658147f7e22))

# [0.6.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.5.1...@querycap-ui/core@0.6.0) (2020-03-30)

### Features

- **ui/search-controls:** added SearchBox and inputs ([072bb9f](https://github.com/querycap/webappkit/commit/072bb9ffb246115a1104c1e38a4278bdf17b1611))

## [0.5.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.5.0...@querycap-ui/core@0.5.1) (2020-03-27)

### Bug Fixes

- **ui/core:** enhance hoisting ([7485b50](https://github.com/querycap/webappkit/commit/7485b50fa8c475cf9150bd3fcc7a0647d4968bc9))

# [0.5.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.4.2...@querycap-ui/core@0.5.0) (2020-03-27)

### Features

- **ui/core:** macro ([a7c90f2](https://github.com/querycap/webappkit/commit/a7c90f266d6338b77ec1a803c75a391bf051017c))
- **ui/core:** renames & currified ([d0bd2ec](https://github.com/querycap/webappkit/commit/d0bd2ec91a2f8ba0a9701c28238fb72fb10430e1))

## [0.4.2](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.4.1...@querycap-ui/core@0.4.2) (2020-03-25)

### Bug Fixes

- dark mode enhance ([10b4685](https://github.com/querycap/webappkit/commit/10b4685758ca6e85379e164dbcacad0df9f99a09))

## [0.4.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.4.0...@querycap-ui/core@0.4.1) (2020-03-25)

**Note:** Version bump only for package @querycap-ui/core

# [0.4.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.3.1...@querycap-ui/core@0.4.0) (2020-03-25)

### Features

- dark mode ([8836b12](https://github.com/querycap/webappkit/commit/8836b12e58ec24e846bcbc14482d8e67b7c5c5da))

## [0.3.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.3.0...@querycap-ui/core@0.3.1) (2020-03-24)

### Bug Fixes

- **form-controls:** styling fix Switch ([16560f1](https://github.com/querycap/webappkit/commit/16560f1865cb19631e6e33957935645cb4abcdef))

# [0.3.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.2.0...@querycap-ui/core@0.3.0) (2020-03-24)

### Features

- **blocks:** add Modal & Popover & Tooltip ([d0cd566](https://github.com/querycap/webappkit/commit/d0cd566f136afe6f29bdab39aeebf55afd04e53d))
- **blocks:** added Dialogs ([bfcc7fe](https://github.com/querycap/webappkit/commit/bfcc7fec2bd8d39d634091780cb3cce7b535d4be))

# 0.2.0 (2020-03-22)

### Features

- **sg:** setup sg ([f612545](https://github.com/querycap/webappkit/commit/f61254522da7bd1bb87c0313036dee2da88cbaf1))
- **ui/core:** rename Theme ([cc8fce5](https://github.com/querycap/webappkit/commit/cc8fce5eea595c2e50c00d0243d2e4c08c3bab9e))
- **uikit:** setup uikit ([7190c50](https://github.com/querycap/webappkit/commit/7190c50a0e5566ea3ef85cba0b1218b4e3d47ba8))

# 0.1.0 (2020-03-13)

### Features

- **ui/core:** rename Theme ([cc8fce5](https://github.com/querycap/webappkit/commit/cc8fce5eea595c2e50c00d0243d2e4c08c3bab9e))
- **uikit:** setup uikit ([7190c50](https://github.com/querycap/webappkit/commit/7190c50a0e5566ea3ef85cba0b1218b4e3d47ba8))
