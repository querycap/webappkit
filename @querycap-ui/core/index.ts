export * from "polished";
export * from "./theme";
export * from "./select";
export * from "./domHelpers";
export * from "react-use-gesture";
export * from "react-spring";
export { keyframes } from "@emotion/react";
