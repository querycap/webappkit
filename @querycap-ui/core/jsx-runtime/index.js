exports.Fragment = require("@emotion/react/jsx-runtime").Fragment;
exports.jsx = require("@emotion/react/jsx-runtime").jsx;
exports.jsxs = require("@emotion/react/jsx-runtime").jsxs;