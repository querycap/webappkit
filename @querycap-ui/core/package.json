{
  "name": "@querycap-ui/core",
  "version": "0.11.6",
  "license": "MIT",
  "peerDependencies": {
    "@babel/runtime-corejs3": "*",
    "lodash": "4.x",
    "react": "17.x"
  },
  "dependencies": {
    "@emotion/react": "^11.4.1",
    "@querycap-ui/core.macro": "workspace:^0.4.4",
    "@querycap-ui/css-aliases": "workspace:^0.3.4",
    "polished": "^4.1.3",
    "react-spring": "^9.2.4",
    "react-use-gesture": "^9.1.3"
  },
  "publishConfig": {
    "access": "public"
  },
  "monobundle": {
    "env": "browser",
    "sideDependencies": [
      "@querycap-ui/core.macro"
    ]
  },
  "scripts": {
    "prepublishOnly": "../../node_modules/.bin/monobundle"
  },
  "files": [
    "jsx-runtime/",
    "macro/",
    "./dist/",
    "dist/"
  ],
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/querycap/webappkit.git"
  },
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.mjs"
    },
    "./jsx-runtime": {
      "require": "./jsx-runtime/index.js",
      "import": "./jsx-runtime/index.mjs"
    },
    "./macro": {
      "require": "./macro/index.js"
    },
    "./index.ts": "./index.ts"
  },
  "types": "./dist/index.d.ts",
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs"
}
