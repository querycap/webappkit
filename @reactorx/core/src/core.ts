import { BehaviorSubject, Observable, Subject } from "rxjs";
import { compose, shallowEqual } from "./utils";

export type TKeyCreator<TActor extends Actor> = (actor: TActor) => string;

export enum AsyncStage {
  STARTED = "STARTED",
  DONE = "DONE",
  FAILED = "FAILED",
  CANCEL = "CANCEL",
}

export interface IActorOpt<TArg = any, TOpts = any> {
  group?: string;
  name?: string;
  stage?: AsyncStage;
  effect?: (state: any, actor: Actor<TArg, TOpts>) => any;
  arg?: TArg;
  opts?: TOpts;
}

export function effectOn<TRoot, TActor extends Actor>(
  keyOrKeyCreator: string | TKeyCreator<TActor>,
  effect: (root: TRoot, actor: TActor) => TRoot | undefined,
) {
  return function (root: any = {}, actor: TActor) {
    const k = typeof keyOrKeyCreator === "function" ? keyOrKeyCreator(actor) : keyOrKeyCreator;

    if (k === "") {
      return root;
    }

    const nextState = effect(root[k], actor);

    const nextRoot: typeof root = {};

    for (const key in root) {
      if (key === k) {
        continue;
      }
      nextRoot[key] = root[key];
    }

    if (typeof nextState !== "undefined") {
      nextRoot[k] = nextState;
    }

    return nextRoot;
  };
}

export class Actor<TArg = any, TOpts = any> {
  static of<TArg = any, TOpts = any>(group: string) {
    return new this<TArg, TOpts>({ group });
  }

  group = "UN_GROUPED";
  name = "UN_NAMED";
  stage?: AsyncStage;

  effect?: (state: any, actor: Actor<TArg, TOpts>) => any;

  arg: TArg = {} as TArg;
  opts: TOpts = {} as TOpts;

  constructor(opt: IActorOpt<TArg, TOpts>) {
    this.group = opt.group || this.group;
    this.name = opt.name || "";
    this.stage = opt.stage;

    this.effect = opt.effect;

    this.arg = opt.arg as TArg;
    this.opts = opt.opts || ({} as TOpts);
  }

  named<TNamedArg = TArg, TNamedOpts = TOpts>(name: string, opts?: TNamedOpts): Actor<TNamedArg, TNamedOpts & TOpts> {
    return new (this.constructor as any)({
      ...(this as any),
      name,
      opts: {
        ...(this.opts as any),
        ...(opts as any),
      },
    });
  }

  effectWith(effect: (state: any, actor: Actor<TArg, TOpts>) => any): this {
    return new (this.constructor as any)({
      ...(this as any),
      effect,
    });
  }

  effectOn<TState = any>(
    keyOrKeyCreator: string | TKeyCreator<Actor<TArg, TOpts>>,
    effect: (state: TState, actor: Actor<TArg, TOpts>) => TState | undefined,
  ): this {
    return this.effectWith(effectOn(keyOrKeyCreator, effect));
  }

  with(arg: TArg, opts?: TOpts): this {
    return new (this.constructor as any)({
      ...(this as any),
      arg,
      opts: {
        ...(this.opts as any),
        ...(opts as any),
      },
    });
  }

  invoke(dispatcher: { dispatch: (actor: Actor) => void }) {
    dispatcher.dispatch(this);
  }

  get type(): string {
    const opts = {} as any;
    let hasOpts = false;

    for (const k in this.opts) {
      if (k !== "parentActor") {
        opts[k] = this.opts[k];
        hasOpts = true;
      }
    }

    return `@@${this.group}/${this.name}${this.stage ? `::${this.stage}` : ""}${
      hasOpts ? `${JSON.stringify(this.opts)}` : ""
    }`;
  }

  is = (actor: Actor): actor is Actor<TArg, TOpts> => {
    const isSame = this.isSameGroup(actor) && actor.name === this.name;
    if (this.stage) {
      return isSame && actor.stage === this.stage;
    }
    return isSame;
  };

  isSameGroup = (actor: Actor<TArg, TOpts>) => actor.group === this.group;
}

export interface IDispatch {
  (actor: Actor): Actor;
}

export interface IMiddlewareAPI<TState = any> {
  dispatch(actor: Actor<TState>): Actor<TState>;

  getState(): TState;
}

export interface IMiddleware<TState = any> {
  (api: IMiddlewareAPI<TState>): (next: IDispatch) => IDispatch;
}

export interface IEpic<TState = any> {
  (actor$: Observable<Actor>, so$: Store<TState>): Observable<Actor>;
}

export function composeEpics(...epics: Array<IEpic<any>>): IEpic<any> {
  return (actor$: Observable<Actor>, so$: Store): Observable<Actor> => {
    return epics.reduce((preActor$, epic) => {
      return epic(preActor$, so$) as any;
    }, actor$);
  };
}

export class Store<TRoot extends { [key: string]: any } = {}> extends BehaviorSubject<TRoot> {
  static create<TState>(initialState: TState = {} as TState) {
    return new Store<TState>(initialState);
  }

  actor$ = new Subject<Actor>();

  applyMiddleware(...middlewares: IMiddleware<TRoot>[]) {
    if (middlewares.length === 0) {
      return;
    }
    const chain = middlewares.map((middleware) => middleware(this));
    this.dispatch = compose(...chain)(this._dispatch);
  }

  epicOn(epic: IEpic<TRoot>) {
    return epic(this.actor$, this).subscribe((actor) => {
      if (actor) {
        this.dispatch(actor);
      }
    });
  }

  dispatch = (actor: Actor) => {
    return this._dispatch(actor);
  };

  getState = (): TRoot => {
    return this.value;
  };

  private _dispatch = (actor: Actor): Actor => {
    if (actor.effect) {
      const nextValue = actor.effect(this.value, actor);
      if (!shallowEqual(nextValue, this.value)) {
        this.next(nextValue);
      }
    }
    this.actor$.next(actor);
    return actor;
  };
}
