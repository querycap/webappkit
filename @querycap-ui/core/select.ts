import { isEmpty, isFunction } from "lodash";
import { Interpolation } from "@emotion/react";
import { CSSProperties } from "react";
import { aliases } from "@querycap-ui/css-aliases";
import { Theme } from "./theme";

export type InterpolationBuilder = (t: Theme) => Interpolation<any>;

const applyStyles =
  (...interpolations: Array<InterpolationBuilder | Interpolation<any>>) =>
  (t: any): Interpolation<any> => {
    const styles: Interpolation<any>[] = [];

    for (const interpolation of interpolations) {
      styles.push(isFunction(interpolation) ? interpolation(t) : interpolation);
    }

    if (styles.length === 0) {
      return {};
    }

    if (styles.length === 1) {
      return styles[0];
    }

    return styles;
  };

export interface CSSPropertiesWithAliases extends CSSProperties {
  colorFill: CSSProperties["color"];
  borderX: CSSProperties["borderLeft"];
  borderY: CSSProperties["borderLeft"];
  marginX: CSSProperties["marginTop"];
  marginY: CSSProperties["marginTop"];
  paddingX: CSSProperties["paddingTop"];
  paddingY: CSSProperties["paddingTop"];
  borderTopRadius: CSSProperties["borderTopLeftRadius"];
  borderBottomRadius: CSSProperties["borderTopLeftRadius"];
  borderLeftRadius: CSSProperties["borderTopLeftRadius"];
  borderRightRadius: CSSProperties["borderTopLeftRadius"];
}

export interface BuilderProperties extends CSSPropertiesWithAliases {
  with: Interpolation<any>;
}

export type CSSBuilder = {
  [k in keyof BuilderProperties]-?: (arg: ((t: Theme) => BuilderProperties[k]) | BuilderProperties[k]) => CSSBuilder;
} & {
  <T extends {}>(t: T): Interpolation<T>;
};

type StyleOrBuilderSet = {
  [k: string]: any | ((t: any) => any);
};

const buildStyle =
  (styleOrBuilders: StyleOrBuilderSet) =>
  (t: any): Interpolation<any> => {
    const styles = {} as any;

    for (const prop in styleOrBuilders) {
      const styleOrBuilder = styleOrBuilders[prop];
      const value = isFunction(styleOrBuilder) ? styleOrBuilder(t) : styleOrBuilder;

      if ((aliases as any)[prop]) {
        for (const p of (aliases as any)[prop]) {
          styles[p] = value;
        }
      } else {
        styles[prop] = value;
      }
    }

    return styles;
  };

export const selectKeys = (...selectors: string[]) => selectors.join(", ");

const createBuilder = (
  selectors: readonly string[],
  styleOrBuilders: StyleOrBuilderSet = {},
  interpolationOrBuilders: ReadonlyArray<InterpolationBuilder | Interpolation<any>> = [],
) => {
  const applyTheme = (t: any) => {
    const n = selectors.length;

    const final = applyStyles(
      ...(isEmpty(styleOrBuilders)
        ? interpolationOrBuilders
        : [...interpolationOrBuilders, buildStyle(styleOrBuilders)]),
    )(t);

    if (n === 0) {
      return final;
    }

    return {
      [selectKeys(...selectors)]: final,
    };
  };

  const builder = new Proxy(applyTheme, {
    get(_, prop) {
      // @emotion 11 should hack return undefined for private prop
      if (prop === "raw" || prop === "__emotion_styles") {
        return undefined;
      }
      if (prop === "with") {
        return (v: any): any => {
          return createBuilder(
            selectors,
            {},
            !isEmpty(styleOrBuilders)
              ? [...interpolationOrBuilders, buildStyle(styleOrBuilders), v]
              : [...interpolationOrBuilders, v],
          );
        };
      }

      return (v: any): any => {
        styleOrBuilders[prop as any] = v;
        return builder;
      };
    },
  }) as CSSBuilder;

  return builder;
};

export function select(...selectors: readonly string[]): CSSBuilder {
  return createBuilder(selectors, {}, []);
}

declare module "react" {
  interface Attributes {
    css?: Interpolation<any>;
  }
}
