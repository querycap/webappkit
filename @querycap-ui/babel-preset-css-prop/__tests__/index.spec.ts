import { transformSync } from "@babel/core";
import presetCSSProp from "@querycap-ui/babel-preset-css-prop";
import { fileURLToPath } from "url";
import { dirname } from "path";

const __dirname = dirname(fileURLToPath(import.meta.url));

const compileToSnapshot = (code: string) => {
  const t = transformSync(code, {
    root: __dirname,
    presets: [presetCSSProp],
  });

  return `
${code}
    
 ↓ ↓ ↓ ↓ ↓ ↓

${t?.code}  
  `;
};

describe("babel-preset-css-prop", () => {
  it("without css prop", () => {
    expect(
      compileToSnapshot(`
const SomeComponent = () => {
  return (
    <>
      <div>123</div>
      <span>{undefined}</span>  
      <span v="v" key={"i"} {...v}  />
    </>
  )
}`),
    ).toMatchSnapshot();
  });

  it("css prop", () => {
    expect(
      compileToSnapshot(`
const SomeComponent = () => {
  return (
    <>
      <div css={(t) => ({ color: t.color.primary })}>123</div>
    </>
  )
}`),
    ).toMatchSnapshot();
  });

  it("css prop only", () => {
    expect(
      compileToSnapshot(`
const SomeComponent = () => {
  return (
    <div css={(t) => ({ color: t.color.primary })}><span key={"a"}>1</span><span key={"b"}>2</span></div>
  )
}`),
    ).toMatchSnapshot();
  });
});
