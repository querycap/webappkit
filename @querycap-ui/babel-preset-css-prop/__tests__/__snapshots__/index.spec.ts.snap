// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-preset-css-prop css prop 1`] = `
"

const SomeComponent = () => {
  return (
    <>
      <div css={(t) => ({ color: t.color.primary })}>123</div>
    </>
  )
}
    
 ↓ ↓ ↓ ↓ ↓ ↓

import { jsx as _jsx } from \\"@querycap-ui/core/jsx-runtime\\";
import { Fragment as _Fragment } from \\"@querycap-ui/core/jsx-runtime\\";

const SomeComponent = () => {
  return _jsx(_Fragment, {
    children: _jsx(\\"div\\", {
      css: t => ({
        color: t.color.primary
      }),
      children: \\"123\\"
    })
  });
};  
  "
`;

exports[`babel-preset-css-prop css prop only 1`] = `
"

const SomeComponent = () => {
  return (
    <div css={(t) => ({ color: t.color.primary })}><span key={\\"a\\"}>1</span><span key={\\"b\\"}>2</span></div>
  )
}
    
 ↓ ↓ ↓ ↓ ↓ ↓

import { jsx as _jsx } from \\"@querycap-ui/core/jsx-runtime\\";
import { jsxs as _jsxs } from \\"@querycap-ui/core/jsx-runtime\\";

const SomeComponent = () => {
  return _jsxs(\\"div\\", {
    css: t => ({
      color: t.color.primary
    }),
    children: [_jsx(\\"span\\", {
      children: \\"1\\"
    }, \\"a\\"), _jsx(\\"span\\", {
      children: \\"2\\"
    }, \\"b\\")]
  });
};  
  "
`;

exports[`babel-preset-css-prop without css prop 1`] = `
"

const SomeComponent = () => {
  return (
    <>
      <div>123</div>
      <span>{undefined}</span>  
      <span v=\\"v\\" key={\\"i\\"} {...v}  />
    </>
  )
}
    
 ↓ ↓ ↓ ↓ ↓ ↓

import { jsx as _jsx } from \\"react/jsx-runtime\\";
import { Fragment as _Fragment } from \\"react/jsx-runtime\\";
import { jsxs as _jsxs } from \\"react/jsx-runtime\\";

const SomeComponent = () => {
  return /*#__PURE__*/_jsxs(_Fragment, {
    children: [/*#__PURE__*/_jsx(\\"div\\", {
      children: \\"123\\"
    }), /*#__PURE__*/_jsx(\\"span\\", {
      children: undefined
    }), /*#__PURE__*/_jsx(\\"span\\", {
      v: \\"v\\",
      ...v
    }, \\"i\\")]
  });
};  
  "
`;
